### PR TITLE
add support for aws assume role credentials

### PR DESCRIPTION
**Pull request by hnaoto**
#### Description
- The existing implementation doesn't support temporary IAM credentials. A user has to create an IAM user, get `iam_access_key` and `iam_secret_key` and then use those values to create a Ruby Kafka client. 
- This MR will add support for aws iam assume role credentials. User will be able to retrieve `AssumeRoleCredentials` from a STS client and use the `AssumeRoleCredentials` to create a Ruby Kafka client in the following fashion.
```
sts = Aws::STS::Client.new
# to get more information about how to create AssumeRoleCredentials
# please refer to https://docs.aws.amazon.com/sdk-for-ruby/v3/api/Aws/AssumeRoleCredentials.html
role_credentials = Aws::AssumeRoleCredentials.new(
  client: sts,
  role_arn: "role_arn",
  role_session_name: "role_session_name"
  
)
kafka_client = Kafka.new(
  ["broker_address"],
  aws_iam_assume_role_credentials: role_credentials,
  sasl_aws_msk_iam_aws_region: "us-east-2",
  ssl_ca_certs_from_system: true,
)
```

#### Implementation Details
- Although the `AssumeRoleCredentials`  are temporary credentials and will expire in 1 hour by default. The code change in this MR won't need to refresh `AssumeRoleCredentials`.
- The `AssumeRoleCredentials` will be refreshed by AWS SDK in the background automatically (For more details, please refer to https://github.com/aws/aws-sdk-ruby/blob/version-3/gems/aws-sdk-core/lib/aws-sdk-core/refreshing_credentials.rb). The  `AssumeRoleCredentials` is passed to the Kafka client as an object and it will get updated in the background as well.
-  Either `assume_role_credentials` object or the combination of `access_key_id` and `secret_key_id` has to be supplied. If neither of those parameters are supplied, an `AwsMskIamCredentialsException` error will be raised internally and an error message will be logged. The Kafka client will crash eventually due to `ConnectionError`. 

#### Discussions 
Discussions regarding implementation details and testing can be found in https://github.com/zendesk/ruby-kafka/issues/944

#### Documentations and Comments
- [x] add comments
- [x] add descriptions to readme
- [x] add sample code to example directory

#### Testing
- [x] test with a small MSK cluster and a skeleton ruby program...
- [x] add unit tests
- [x] test in working environment

### CODE REVIEW

**Comment by hnaoto** [comment]
Hi @garrett528 and @jobeus, would you mind taking a look at the MR?
Please let me know if there is anything appears to be unclear to you.

(I will try to loop in the maintainers of this repo later.

File: lib/kafka/sasl/awsmskiam.rb
Comment (garrett528) [inline lib/kafka/sasl/awsmskiam.rb:141]: does this get messed up if the session token isn't passed and defaults to `nil`?

File: lib/kafka/sasl/awsmskiam.rb
Comment (garrett528) [inline lib/kafka/sasl/awsmskiam.rb:120]: see below. it should be acceptable to not pass a session token if you are using permanent credentials. in that case, we don't want to set this parameter (i think it'll fail if we set it to `nil` but want to confirm that that's the behavior).

Comment (hnaoto) [reply lib/kafka/sasl/awsmskiam.rb:120]: Yes, I think you are right.

When I test the existing implementation, I did notice that I have to pass `session_token` when I use static credentials. I didn't realize it was caused by setting 'x-amz-security-token' as `nil`... Let me try to set `x-amz-security-token` as optional

Comment (hnaoto) [reply lib/kafka/sasl/awsmskiam.rb:120]: I think that is a good point. `.delete_if { |_, v| v.nil? }` should be able to remove any keypair if the value is nil

File: lib/kafka/sasl/awsmskiam.rb
Comment (hnaoto) [reply lib/kafka/sasl/awsmskiam.rb:141]: should still work.. `.delete_if { |_, v| v.nil? }` will remove `"X-Amz-Security-Token" => nil`

**Comment by jobeus** [comment]
lgtm in general but changed jobs and don't have access to a kafka cluster in AWS right now to test on, etc..

**Comment by hnaoto** [comment]
Thanks @garrett528 😄!  really appreciate your help. Thanks for joining the discussion and sharing your insights

**Comment by hnaoto** [comment]
Thanks for reviewing the MR @jobeus! 
No worry. I will try to test the code change in a working environment

**Comment by hnaoto** [comment]
Hi @dasch and @leonmaia, would you mind taking a look at this MR?

Please let me know if you need any further information.

**Comment by jordanfbrown** [comment]
Hey @hnaoto, this looks great and is something my company is interested in. Wondering what the next steps are to get this approved and merged? I'm happy to help out with testing/finalizing the PR/etc.

**Comment by hnaoto** [comment]
Hi @jordanfbrown, if you are interested in testing the MR, please give it a try. If you encounter any issue, please let me know. really appreciate that.

Please noted that MSK currently has connection constraints when one uses IAM authentication. If a large number of Kafka clients trying to establish connections through IAM authentication at the same time, there will be some serious connection issues. However, it depends on the number of the connections that one attempts to establish simultaneously, and your clients could work perfectly fine if they don't hit the bottleneck.


To get the MR merged we need to have it reviewed by @dasch.

**Comment by dasch** [comment]
I would love to see users try this out before merging.

File: README.md
Comment (dasch) [inline README.md:1138]: Can you make some of these terms links to the relevant AWS docs? First time users would benefit from that.

**Review by dasch — APPROVED** [review]
Other than that minor comment I'm :+1:

Would you like to get more testing by other users, or do you think this is stable?

Finally, can you update the CHANGELOG with this change?

**Pull request closed without merging**
